ddtrace/propagation/http.py

```diff
@@ -853,16 +857,45 @@ def my_controller(url, headers):
         """
         if not headers:
             return Context()
-
         try:
             normalized_headers = {name.lower(): v for name, v in headers.items()}
 
-            # loop through the extract propagation styles specified in order
-            for prop_style in config._propagation_style_extract:
-                propagator = _PROP_STYLES[prop_style]
-                context = propagator._extract(normalized_headers)  # type: ignore
-                if context is not None:
-                    return context
+            # or tracer configured to extract first only
+            if config._propagation_extract_first:
+                # loop through the extract propagation styles specified in order, return whatever context we get first
+                for prop_style in config._propagation_style_extract:
+                    propagator = _PROP_STYLES[prop_style]
+                    context = propagator._extract(normalized_headers)  # type: ignore
+                    if context is not None:
+                        return context
+            # loop through all extract propagation styles
```

Comment:
You could reduce the complexity and depth of this nested conditional by factoring the branches into functions that do what these comments say.